### PR TITLE
Rename the `TrainingTypePlugin` base to `Strategy`

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Renames the base class to `Strategy`. 
The next PRs will cover all related usages:
- Renaming subclasses, e.g., DDPPlugin -> DDPStrategy
- module training_type -> strategy
- Registry: #11122
- other variable names across the project

Part of #10549

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
I made sure I had fun coding 🙃

Part of #1 (it's a lie, this is just here to avoid noisy GitHub bot)

cc @borda @justusschock @awaelchli @akihironitta

### CODE REVIEW

**Review by carmocca — APPROVED** [review]

**Review by justusschock — APPROVED** [review]

File: docs/source/extensions/plugins.rst [hunk 1]
```diff
@@ -29,7 +29,7 @@ We expose Accelerators and Plugins mainly for expert users that want to extend L
 
 There are two types of Plugins in Lightning with different responsibilities:
 
-TrainingTypePlugin
+Strategy
 ------------------
```
Comment (rohitgr7) [inline docs/source/extensions/plugins.rst:33]:
```suggestion
Strategy
--------
```

File: pytorch_lightning/plugins/training_type/training_type_plugin.py [hunk 2]
```diff
@@ -37,7 +37,7 @@
 TBroadcast = TypeVar("TBroadcast")
 
 
-class TrainingTypePlugin(ABC):
+class Strategy(ABC):
```
Comment (rohitgr7) [inline pytorch_lightning/plugins/training_type/training_type_plugin.py:40]: the filename will be renamed later?

also, this isn't going to follow a deprecation process?

Comment (justusschock) [reply pytorch_lightning/plugins/training_type/training_type_plugin.py:40]: This was still marked as experimental, not sure if we need one here

Comment (awaelchli) [reply pytorch_lightning/plugins/training_type/training_type_plugin.py:40]: Yes, (my) plan is to rename components first, then the files. 
Happy to hear suggestions how to "deprecate" things, but as Justus said it is inevitable that 1.6 will break people's custom plugins unfortunately, considering all the changes we already did. 

But we can deprecate things on an import level for example. People importing from `plugins.training_type` should get a warning that it was moved to `strategy.x`, maybe?

Comment (rohitgr7) [reply pytorch_lightning/plugins/training_type/training_type_plugin.py:40]: if it's experimental then it's fine.
is it targeted to stabilize by v1.6?

**Review by rohitgr7 — APPROVED** [review]

File: pytorch_lightning/plugins/training_type/training_type_plugin.py (hunk 2 above)
Comment (awaelchli) [reply pytorch_lightning/plugins/training_type/training_type_plugin.py:40]: 100%

**Comment by four4fish** [comment]
I marked this as breaking change as user who extend TrainingTypePlugin will need to update their class inheritance

File: pytorch_lightning/trainer/trainer.py [hunk 3]
```diff
@@ -1681,7 +1681,7 @@ def accelerator(self) -> Accelerator:
         return self.training_type_plugin.accelerator
 
     @property
-    def training_type_plugin(self) -> TrainingTypePlugin:
+    def training_type_plugin(self) -> Strategy:
```
Comment (four4fish) [inline pytorch_lightning/trainer/trainer.py:1684]: Are we planning on completely get rid of training_type_plugin this word in code base? It still heavily used in trainer.py and accelerator_connector.py

**Review by four4fish — APPROVED** [review]

File: pytorch_lightning/plugins/training_type/training_type_plugin.py (hunk 2 above)
Comment (Borda) [reply pytorch_lightning/plugins/training_type/training_type_plugin.py:40]: > Yes, our plan is to rename components first, then the files.

I would rather make this in single PR, as it trivial rename shall be fine

**Review by Borda — APPROVED** [review]

File: pytorch_lightning/trainer/trainer.py (hunk 3 above)
Comment (awaelchli) [reply pytorch_lightning/trainer/trainer.py:1684]: Yes, everywhere. It only makes sense if it is consistent everywhere.

File: pytorch_lightning/plugins/training_type/training_type_plugin.py (hunk 2 above)
Comment (awaelchli) [reply pytorch_lightning/plugins/training_type/training_type_plugin.py:40]: Unfortunately this will result in 100+ files changed, nobody would review this. I would prefer to take it in smaller pieces.

**Review by tchaton — APPROVED** [review]
LGTM !

**Review by kaushikb11 — CHANGES_REQUESTED** [review]
This is a heavy breaking change on the import level itself for users who extend TrainingTypePlugin. Can’t we prevent it by introducing `strategy.py` file in the PR?

**Comment by justusschock** [comment]
@kaushikb11 But then we only delay the breaking changes. I think we could introduce a `training_type_plugin` file for BC, but I really think we should go with these changes as they were considered to be experimental. And since we want to have it stable for 1.6, we cannot delay the BC beyond this

**Comment by awaelchli** [comment]
@kaushikb11 To prevent a huge PR, I would prefer to go step by step by renaming parts individually. Once we move the files we can still add a simple reroute/deprecation path for mapping the training_type module to strategy

**Review by kaushikb11 — APPROVED** [review]
